assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java

```diff
@@ -4656,17 +4655,31 @@ private void notificationToInstructors(AssignmentSubmission submission, Assignme
         }
     }
 
-    private void notificationToStudent(AssignmentSubmission submission) {
+    private void notificationToStudent(AssignmentSubmission submission, Assignment assignment) {
         if (serverConfigurationService.getBoolean("assignment.submission.confirmation.email", true)) {
-            Set<String> submitterIds = submission.getSubmitters().stream().map(AssignmentSubmissionSubmitter::getSubmitter).collect(Collectors.toSet());
-            Set<User> users = submitterIds.stream().map(id -> {
-                try {
-                    return userDirectoryService.getUser(id);
-                } catch (UserNotDefinedException e) {
-                    log.warn("Could not find user with id = {}, {}", id, e.getMessage());
-                }
-                return null;
-            }).filter(Objects::nonNull).collect(Collectors.toSet());
+            String siteId = assignment.getContext();
+            Set<String> siteUsers = null;
+            try {
+                siteUsers = siteService.getSite(siteId).getUsers();
+            } catch (IdUnusedException e) {
+                log.warn("Could not find site: {}", siteId, e);
+            }
+            Set<User> users = submission.getSubmitters()
+                    .stream()
+                    .map(AssignmentSubmissionSubmitter::getSubmitter)
+                    .filter(siteUsers::contains) // this filters active users
```

Comment:
this is the only major change in the PR

Reply:
Paul just pointed this out to me: siteUsers could still be `null` here in the event that an exception is generated in the above try/catch

Reply:
thanks